### PR TITLE
[openload] rewrite extractor

### PR DESCRIPTION
**Pull request by Tithen-Firion**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Rewritten extractor uses [PhantomJS](http://phantomjs.org/) to decode video URL. 

Tests by default fail on Travis because it has PhantomJS version 1.9.8 installed. When running on Trusty (`dist: trusty` in `.travis.yml`) there is PhantomJS 2.0.0. It works but MD5 hash mismatch. I've tested two files: one downloaded with PhantomJS method (fails tests) and another with pairing method (#12603, passes tests) and they are the same (identical MD5 of full file). So IDK why test fails...

### CODE REVIEW

File: youtube_dl/extractor/openload.py [hunk 1]
```diff
@@ -58,68 +62,88 @@ class OpenloadIE(InfoExtractor):
...
+        with script_file:
+            script_file.write(self._PHANTOMJS_SCRIPT % video_id)
+
+        self.to_screen('%s: Decoding video ID with PhantomJS' % video_id)
+
+        p = subprocess.Popen([exe, '--ssl-protocol=any', script_file.name],
+            stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        output, err = p.communicate()
+        if p.returncode != 0:
+            raise ExtractorError('Decoding failed\n:'
```
Comment (dstftw) [inline youtube_dl/extractor/openload.py]: Temp file is not removed in this case.

File: youtube_dl/extractor/openload.py [hunk 2]
```diff
@@ -58,68 +62,88 @@ class OpenloadIE(InfoExtractor):
...
+
     @staticmethod
     def _extract_urls(webpage):
         return re.findall(
             r'<iframe[^>]+src=["\']((?:https?://)?(?:openload\.(?:co|io)|oload\.tv)/embed/[a-zA-Z0-9-_]+)',
             webpage)
 
     def _real_extract(self, url):
+        exe = check_executable('phantomjs', ['-v'])
+        if not exe:
```
Comment (dstftw) [inline youtube_dl/extractor/openload.py]: I would prefer hiding all phantomjs related code in a separate wrapper class.

**Comment by dstftw** [comment]
Have you also considered any other similar to phantomjs solutions?

File: youtube_dl/extractor/openload.py [hunk 3]
```diff
@@ -58,68 +62,88 @@ class OpenloadIE(InfoExtractor):
...
-            r'<span[^>]+class=["\']title["\'][^>]*>([^<]+)', webpage,
-            'title', default=None) or self._html_search_meta(
-            'description', webpage, 'title', fatal=True)
+            raise ExtractorError('File not found', expected=True, video_id=video_id)
+
+        script_file = tempfile.NamedTemporaryFile(mode='w', delete=False)
+
+        # write JS script to file and close it
+        with script_file:
+            script_file.write(self._PHANTOMJS_SCRIPT % video_id)
```
Comment (yan12125) [inline youtube_dl/extractor/openload.py]: The current working directory is not always writable.

Comment (Tithen-Firion) [inline youtube_dl/extractor/openload.py]: `tempfile` writes in temp directory.

Comment (yan12125) [inline youtube_dl/extractor/openload.py]: Sorry - dismiss that

File: youtube_dl/extractor/openload.py (hunk 2 above)
Comment (Tithen-Firion) [inline youtube_dl/extractor/openload.py]: I'm not sure where to put that code. Access to cookie jar would be nice but I don't know where it is.

**Comment by Tithen-Firion** [comment]
@dstftw I have also considered using `jsdom` for `nodejs` but it failed some checks on Openload. Using [Selenium WebDriver](https://pypi.python.org/pypi/selenium) + some browser works too but it's very easy to detect and block them by website. The same case with PhantomJS, it's detectable but in my code I use 
```javascript
page.onInitialized = function() {
  page.evaluate(function() {
    delete window._phantom;
    delete window.callPhantom;
  });
};
```
Haven't tested it but that probably makes some features unavailable. It prevents detecting PhantomJS though.

**Comment by yan12125** [comment]
> It works but MD5 hash mismatch

A possibility: the default User-Agent leads to nice pigeons.

**Comment by Tithen-Firion** [comment]
@yan12125 You're right. I'll add custom headers to download info dict in extractor. But still, it will fail on default Travis build.

**Comment by MemoryAndDream** [comment]
I find it needs "page.settings.userAgent = 'Mozilla/4.0 (compatible; MSIE 6.1; Windows XP)';" while using PHANTOMJS to extract openload

**Comment by Tithen-Firion** [comment]
@dstftw done.

**Comment by mwalcott3** [comment]
> Haven't tested it but that probably makes some features unavailable. It prevents detecting PhantomJS though.
>
It is commonly used to advert this method of detection and doesn't really cause any problems. It only deletes the reference used for the following feature: http://phantomjs.org/api/webpage/handler/on-callback.html. As long as you don't call either of those things then there should be no problem. If you need to use them then you can store the references under a different name.

**Comment by Tithen-Firion** [comment]
Good to know. This could be useful and convenient but not necessary. We can set interval/timeout and check if something is done on the website. I even added the example.

**Comment by yan12125** [comment]
The Travis CI is still falling. Any news for that?

**Comment by Tithen-Firion** [comment]
It should work fine with [`dist: trusty`](https://docs.travis-ci.com/user/trusty-ci-environment/) in `.travis.yml` but it's still in beta so I don't know if you want to use it.

**Comment by yan12125** [comment]
I've used trusty in other projects for a while, so I think it's OK to move on. Could you open another PR and see whether moving to trusty breaks existing tests or not? If trusty is not stable enough, maybe a PPA repo with up-to-date PhantomJS can be used.

Also, if an older PhantomJS does not function, it's better to detect it and give a warning, just like what youtube-dl does for old avconv.

**Comment by Tithen-Firion** [comment]
Okay, I have no idea what's happening. This 2nd PR still uses Precise build and just downloads PhantomJS 2.0 but it still affects tests. Some errors change to failures/successes or the other way around.

**Comment by Tithen-Firion** [comment]
@yan12125 version checking added.

**Comment by yan12125** [comment]
Which PhantomJS version are you using in local tests? In https://travis-ci.org/rg3/youtube-dl/jobs/227100362, youtube-dl attempts to download a wrong URL, and 2.1.1 works fine on my machine. I guess maybe 2.1+ is necessary.

**Comment by Tithen-Firion** [comment]
I haven't applied commits from here to those two PRs. Just wanted to test difference in errors/failures and time. I'll do it now.

**Comment by Hrxn** [comment]
Possible alternative to PhantomJS:

Headless Chrome/Chromium
https://developers.google.com/web/updates/2017/04/headless-chrome

**Comment by Tithen-Firion** [comment]
@Hrxn any example of how to run it from Python *without* Selenium?

**Comment by alexcopy** [comment]
here was an example how to run from node
https://objectpartners.com/2017/04/13/how-to-install-and-use-headless-chrome-on-osx/

**Comment by Tithen-Firion** [comment]
@alexcopy how is installing Node, `chrome-remote-interface` for Node and Chrome Headless better than installing just PhantomJS?

**Comment by Hrxn** [comment]
That would be worse, obviously.

The thing is, you don't really need anything for Headless Chrome. That's just standard Chrome (v59 and onward) that probably anyone has. You can run Headless mode from the command-line.

I don't know about anything specific in regard to Python yet. But I bet there will be Python projects that make use of it.

I am not sure how this stacks up against PhantomJS feature-wise currently, but I guess it will be there soon, probably replacing PhantomJS in the near future for automated website testing, etc.

But I didn't really look into it so far..

Depends on how PhantomJS is currently used in youtube-dl, or rather planned to be used..

**Comment by alexcopy** [comment]
@Tithen-Firion   I managed to run chrome in headless mode, works OK, from command line could run a js and eventually get video url. But for some reasons wasn't been able do the same with PhantomJS, not sure why, possibly content owner has blocked PhantomJS. It's looks like problem in cookies.

**Comment by Tithen-Firion** [comment]
@Hrxn @alexcopy Chrome headless is still [not available](https://bugs.chromium.org/p/chromium/issues/detail?id=686608) on Windows.

@alexcopy [works fine for me](https://gist.github.com/Tithen-Firion/8b3921d745131837519d5c5b95b86440).

**Comment by alexcopy** [comment]
@Tithen-Firion  It's woks fine !! Thx !

**Comment by Hrxn** [comment]
Yes I'm aware, the linked page states that:


Not sure how soon is soon, Chrome Headless for Windows probably will arrive with M60.

I just wanted to mention that there are possible alternatives.

**Comment by gittubes** [comment]
I applied the changes and installed PhantomJS but I keep getting this error:

This is the openload.py file:
https://pastebin.com/LZUiepx7

[Openload] Gc-h_YZTOF4: Downloading webpage
[Openload] Gc-h_YZTOF4: Decoding video ID with PhantomJS
Traceback (most recent call last):
  File "/usr/lib64/python2.7/runpy.py", line 162, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/lib64/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "./youtube-dl/__main__.py", line 19, in <module>
  File "./youtube-dl/youtube_dl/__init__.py", line 465, in main
  File "./youtube-dl/youtube_dl/__init__.py", line 455, in _real_main
  File "./youtube-dl/youtube_dl/YoutubeDL.py", line 1927, in download
  File "./youtube-dl/youtube_dl/YoutubeDL.py", line 762, in extract_info
  File "./youtube-dl/youtube_dl/extractor/common.py", line 433, in extract
  File "./youtube-dl/youtube_dl/extractor/openload.py", line 133, in _real_extract
  File "./youtube-dl/youtube_dl/utils.py", line 654, in encodeArgument
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc2 in position 0: ordinal not in range(128)

**Comment by Tithen-Firion** [comment]
@gittubes you applied just the first commit. Apply all of them.

**Comment by gittubes** [comment]
I thought that there are 3 files that need to be changed right?
Those are the files I changed:

common.py
https://pastebin.com/NizVaFhX
openload.py
https://pastebin.com/LZUiepx7
utils.py
https://pastebin.com/iU0PsgKG

**Comment by Tithen-Firion** [comment]
@gittubes yes, 3 files are changed but you're not using the latest [**openload.py**](https://raw.githubusercontent.com/Tithen-Firion/youtube-dl/c89267d31ad99eb5b1a87cd354de5280a2a087b1/youtube_dl/extractor/openload.py) version.

**Comment by gittubes** [comment]
@Tithen-Firion  Oh, my mistake. Thank you for your support!

**Comment by pelovac** [comment]
I've been using this for a while now, but it no longer works. I keep getting the following error:

ERROR: unable to download video data: <urlopen error no host given>

**Comment by Tithen-Firion** [comment]
@pelovac still works fine for me. Post your PhantomJS version and video URL that caused the error.

**Comment by pelovac** [comment]
PhantomJS version 2.1.1
Openload link: https://openload.co/f/vEgDQukXmrs/

Sometimes I get that error, and sometimes it download a 36Byte file that have the follow:

{"status":500,"msg":"unknown error"}

**Comment by Tithen-Firion** [comment]
@pelovac try running `youtube-dl` with `--verbose` flag, it still works fine for me.

**Comment by Tithen-Firion** [comment]
@dstftw what exactly did you merge? I can't see any changes from this PR in master branch.

**Comment by revolter** [comment]
And it still downloads the pigeons with the latest version.

**Comment by yan12125** [comment]
That might be a Github bug. As "merged" pull requests can't be reopen, could you open a new one? Travis CI now defaults to Ubuntu Trusty, so I guess we can go ahead safely.

**Pull request merged**